### ISSUE TITLE
generalized_lambert_w won't provide float solution

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#619]
Imported from SourceForge on 2024-07-03 03:00:53
Created by **[chrisgorman](https://sourceforge.net/u/chrisgorman/)** on 2023-07-13 00:25:47
Original: https://sourceforge.net/p/maxima/bugs/4156

---

When trying to compute a generalized_lambert_w solution for

~~~
98*t + 1080*%e^(-t/10) - 1180
~~~

I need to

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:00:54
Created by **[chrisgorman](https://sourceforge.net/u/chrisgorman/)** on 2023-07-13 00:39:30
Original: https://sourceforge.net/p/maxima/bugs/4156/#5494

---

Let me try to finish my bug report here...

Sorry for the confusion.

When trying to compute a generalized_lambert_w solution for ...

~~~
98*t + 1080*%e^(-t/10) - 1180
~~~

I need to pass numer or use a float value for z to get a float solution.  For example ...

~~~
590/49 + 10*generalized_lambert_w(-1, (-54/49)*exp(-59/49));
~~~

doesn't work, but either ...

~~~
590/49 + 10*generalized_lambert_w(-1, float((-54/49)*exp(-59/49)));
~~~

or ...

~~~
590/49 + 10*generalized_lambert_w(-1, (-54/49)*exp(-59/49)),numer;
~~~

do work to find a float solution for Wm(z) for my example function.

Perhaps a note can be added to the user documentation describing how to properly manage this use case.

~~~
Maxima-version: "5.47.0"
Maxima build date: "2023-06-19 09:51:41"
Host type: "x86_64-unknown-linux-gnu"
Lisp implementation type: "SBCL"
Lisp implementation version: "2.3.5"
User dir: "/home/chris/src/maxima"
Temp dir: "/home/chris/src/maxima"
Object dir: "/home/chris/.maxima/binary/5_47_0/sbcl/2_3_5"
Frontend: false
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:00:58
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-07-13 02:29:20
Original: https://sourceforge.net/p/maxima/bugs/4156/#ae6f

---

This works for me:
```
generalized_lambert_w(-1,-((54*%e^-(59/49))/49)), numer;
```
But I think
```
float(generalized_lambert_w(-1,-((54*%e^-(59/49))/49)));
```
but it doesn't.  Probably because the first arg is -1.0, and not -1.  I suspect we could modify the simplifier to treat -1.0 (or any other float that is equal to an integer) as -1.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:01
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-07-13 14:22:45
Original: https://sourceforge.net/p/maxima/bugs/4156/#590a

---

Not sure I understand the problem.
Maxima always preserves exact symbolic expressions.
It only calculates floating-point approximations for floating-point arguments or when simplifying with the **numer** flag set.
This is not specific to the Lambert function:
~~~~
sin(1) => sin(1)
sin(1.0) => 0.84...
block([numer:true],sin(1)] => 0.84
~~~~
So I'm not sure what the problem is with Lambert.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:05
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-07-25 14:20:54
Original: https://sourceforge.net/p/maxima/bugs/4156/#590a/154a

---

What's happening is that `float(generalized_lambert_w(-1,x))` becomes `generalized_lambert_w(-1.0, float(x))` and  the simplifier for `generalized_lambert_w` checks for the fixnum  -1 so -1.0 is wrong.  I think the fix should be straightforward.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:12
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-07-25 17:42:38
Original: https://sourceforge.net/p/maxima/bugs/4156/#590a/154a/c123/fbce

---

Chris, no, you shouldn't have to use numer. I think Ray is fixing this problem.
Strictly speaking, I suppose generalized_lambert_w(<float>, ...) is wrong in some sense, since the first argument must be an integer, but it seems OK to accept a float which is an exact integer.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:15
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-07-25 21:35:50
Original: https://sourceforge.net/p/maxima/bugs/4156/#590a/154a/c123/fbce/c232

---

Yeah, this is just the way float works.  It floats all the args and then calls the function again.  Short of making float smarter to handle this function in a special way, I'm making generalized_lambert_w allow float/bfloat values that are integers to be treated as if they were integers.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:19
Created by **[chrisgorman](https://sourceforge.net/u/chrisgorman/)** on 2023-07-25 22:33:10
Original: https://sourceforge.net/p/maxima/bugs/4156/#590a/154a/c123/fbce/c232/aefc

---

Hi Raymond,

Great!  Thanks.

Chris

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:22
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-07-25 23:33:19
Original: https://sourceforge.net/p/maxima/bugs/4156/#6cd4

---

- **status**: open --> closed
- **assigned_to**: David Billinghurst --> Raymond Toy

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 03:01:26
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-07-25 23:33:20
Original: https://sourceforge.net/p/maxima/bugs/4156/#5508

---

Fixed in commit [c293301]

Basically test if the k arg is numerically equal to an integer and use the integer value instead of the float value.
